seagull/lifeforms/wiki.py

```diff
@@ -209,3 +240,157 @@ def parse_cells(cells_str: str) -> Lifeform:
+
+    Args:
+        rle_str (str): single line of RLE commands
+
+    Returns:
+        str: valid PlainText-encoded lifeform
+    """
+    # drop the last part
+    if "!" in rle_str:
+        rle_str = rle_str[: rle_str.index("!")]
+    else:
+        raise ValueError('Incorrect input: no "!"')
+
+    if not set(rle_str).issubset("0123456789bo$"):
+        raise ValueError("Incorrect input: wrong character set")
+
+    commands = re.findall("([0-9]*)(b|o|\$)", rle_str)
+    if len(commands) == 0:
+        raise ValueError("Incorrect input: wrong pattern format")
+
+    layout_string = ""
+    parse_dict = {"b": ".", "o": "O", "$": "\n"}
+    for com in commands:
+        n = int(com[0]) if com[0] else 1
+        layout_string += parse_dict[com[1]] * n
+
+    return layout_string
+
+
+def parse_rle(rle_str: str) -> Lifeform:
+    """Parse rle_str, stored in RLE format, into Lifeform
+    
+    RLE format description: https://www.conwaylife.com/wiki/Run_Length_Encoded
+
+    Usage
+    -----
+    You can enter a string directly into the function: 
+
+    G = parse_rle(
+        '''#N Gosper glider gun
+#O Bill Gosper
+#C A true period 30 glider gun.
+#C The first known gun and the first known finite pattern with unbounded growth.
+#C www.conwaylife.com/wiki/index.php?title=Gosper_glider_gun
+x = 36, y = 9, rule = B3/S23
+24bo11b$22bobo11b$12b2o6b2o12b2o$11bo3bo4b2o12b2o$2o8bo5bo3b2o14b$2o8b
+o3bob2o4bobo11b$10bo5bo7bo11b$11bo3bo20b$12b2o!
+'''
+        )
+    
+    Or you can parse rle files immediately from Conway Life's website:
+
+    G = parse_rle('http://www.conwaylife.com/patterns/gosperglidergun.rle')
+
+
+    Parameters
+    ----------
+    rle_str : str
+        RLE encoded lifeform description
+        May be a filename or a URL to be (down)loaded from        
+
+    Raises
+    ------
+    ValueError
+        if invalid input provided
+
+    Notes
+    -----
+        - RLE content after `!` is ignored    
+        - header line is currently parsed but not used
+    """
+    if not rle_str.startswith(("#", "x")):
+        # not a proper .cells line, filename/URL?
+        rle_str = _load_file_of_url(rle_str)
+
+    # split lines, \r if (down)loaded and not copy-pasted
+    lines = rle_str.strip().replace("\r\n", "\n").split("\n")
+
+    metadata_lines = [l for l in lines if l.startswith("#")]
+    layout_lines = [l for l in lines if not l.startswith("#")]
+
+    # Parse size and rule, if present
+    header_line = layout_lines[0]
+    header_match = re.match(
+        r"x = ([0-9]+), y = ([0-9]+)(, rule = ([^ ]+))?", header_line
+    )
+    if header_match is None:
+        raise ValueError(
+            f"Incorrect input: wrong header line format : [{header_line}]"
+        )
+    header_match = header_match.groups()
+
+    # Hm.. do these really matter if I do not parse #R,#P,#r comments?
+    #  possible usage: check dimension of parser plaintext layout
+    #   which could differ really if rle is not optimized
```

Comment:
I like what you did above where you put `@TODO` and `@FIXME` for things you're not sure or can still be optimized. I suggest doing the same thing here :+1: